synthesizer/inference.py

```diff
@@ -87,37 +84,38 @@ def synthesize_spectrograms(self, texts: List[str],
         :return: a list of N melspectrograms as numpy arrays of shape (80, Mi), where Mi is the 
         sequence length of spectrogram i, and possibly the alignments.
         """
-        if not self._low_mem:
-            # Usual inference mode: load the model on the first request and keep it loaded.
-            # Reload it every time for deterministic operation if seed specified.
-            if not self.is_loaded() or self._seed is not None:
-                self.load()
-            specs, alignments = self._model.my_synthesize(embeddings, texts)
-        else:
-            # Low memory inference mode: load the model upon every request. The model has to be 
-            # loaded in a separate process to be able to release GPU memory (a simple workaround 
-            # to tensorflow's intricacies)
-            specs, alignments = Pool(1).starmap(Synthesizer._one_shot_synthesize_spectrograms, 
-                                                [(self.checkpoint_fpath, embeddings, texts)])[0]
-    
-        return (specs, alignments) if return_alignments else specs
+        # Load the model on the first request. For low_mem it is loaded every time.
+        if not self.is_loaded():
+            self.load()
 
-    @staticmethod
-    def _one_shot_synthesize_spectrograms(checkpoint_fpath, embeddings, texts):
-        # Load the model and forward the inputs
-        tf.compat.v1.reset_default_graph()
-        model = Tacotron2(checkpoint_fpath, hparams, seed=self._seed)
-        specs, alignments = model.my_synthesize(embeddings, texts)
-        
-        # Detach the outputs (not doing so will cause the process to hang)
-        specs, alignments = [spec.copy() for spec in specs], alignments.copy()
-        
-        # Close cuda for this process
-        model.session.close()
-        numba.cuda.select_device(0)
-        numba.cuda.close()
-        
-        return specs, alignments
+            # Print some info about the model when it is loaded            
+            tts_k = self._model.get_step() // 1000
+
+            simple_table([("Tacotron", str(tts_k) + "k"),
+                        ("r", self._model.r)])
+
+
+        inputs = [text_to_sequence(text.strip(), hparams.tts_cleaner_names) for text in texts]
+        if not isinstance(embeddings, list):
+            embeddings = [embeddings]
+
+        specs = []
+        for i, x in enumerate(inputs, 1):
+
+            print(f"\n| Generating {i}/{len(inputs)}")
+
+            speaker_embedding = torch.tensor(embeddings[i-1]).float().to(self.device)
+            _, m, alignments = self._model.generate(x, speaker_embedding)
+            specs.append(m)
```

Comment:
Why's it feeding inputs one by one instead of leveraging batching?

Reply:
Short answer: That's how fatchord does it.

Longer answer: I tried to avoid making unnecessary changes to fatchord's code, figuring that cleanups and performance enhancements could come later. I implemented batching and didn't push it when I found that it didn't speed up inference. However, that was on CPU and I need to retest on GPU.

The synthesizer is noticeably slower in pytorch than tensorflow. Haven't formally benchmarked it but the existing TF model is about 2-3x faster for longer, multi-line inputs.

Reply:
Added input batching for inference in https://github.com/CorentinJ/Real-Time-Voice-Cloning/pull/472/commits/0ba3ba36dba44d000598b17687bd54ca8de52721

The resulting code is not as readable, but it might be faster.